src/goto-instrument/contracts/dynamic-frames/dfcc_instrument.cpp

```diff
@@ -0,0 +1,1359 @@
+
+    insert_before_swap_and_advance(goto_program, target, payload);
+
+    // step back
+    target--;
+  }
+}
+
+#if 0
+// TODO re-enable once we can have nondet-static with with arbitrary LHS
+// expressions in the INITIALIZE_FUNCTION function
+void dfcc_instrumentt::instrument_fptr_call_instruction_dynamic_lookup(
+  const exprt &write_set,
+  goto_programt::targett &target,
+  goto_programt &goto_program)
+{
+  // Insert a dynamic lookup in __instrumented_functions_map
+  // and pass the write set only to functions that are known to be able
+  // to accept it.
+  //
+  // ```
+  // IF __instrumented_functions_map[__CPROVER_POINTER_OBJECT(fptr)] != 1
+  //   GOTO no_inst;
+  // CALL [lhs] = fptr(params, write_set);
+  // GOTO end;
+  // no_inst:
+  // CALL [lhs] = fptr(params);
+  // end:
+  // SKIP;
+  // ---
+  // SKIP // [lhs] = fptr(params) turned into SKIP
+  // ```
+  const auto &target_location = target->source_location();
+  const auto &callf = target->call_function();
+  auto object_id = pointer_object(
+    (callf.id() == ID_dereference) ? to_dereference_expr(callf).pointer()
+                                   : address_of_exprt(callf));
+  auto index_expr = index_exprt(
+    library.get_instrumented_functions_map_symbol().symbol_expr(), object_id);
+  auto cond = notequal_exprt(index_expr, from_integer(1, c_bool_typet(8)));
+  goto_programt payload;
+  auto goto_no_inst =
+    payload.add(goto_programt::make_incomplete_goto(cond, target_location));
+  code_function_callt call_inst(
+    target->call_lhs(), target->call_function(), target->call_arguments());
+  call_inst.arguments().push_back(write_set);
+  payload.add(goto_programt::make_function_call(call_inst, target_location));
+  auto goto_end_inst = payload.add(
+    goto_programt::make_incomplete_goto(true_exprt(), target_location));
+  auto no_inst_label = payload.add(goto_programt::make_skip(target_location));
+  goto_no_inst->complete_goto(no_inst_label);
+  code_function_callt call_no_inst(
+    target->call_lhs(), target->call_function(), target->call_arguments());
+  payload.add(goto_programt::make_function_call(call_no_inst, target_location));
+  auto end_label = payload.add(goto_programt::make_skip(target_location));
+  goto_end_inst->complete_goto(end_label);
+  // erase the original call
+  target->turn_into_skip();
+  insert_before_swap_and_advance(goto_program, target, payload);
+}
+#endif
+void dfcc_instrumentt::instrument_call_instruction(
+  const exprt &write_set,
+  goto_programt::targett &target,
+  goto_programt &goto_program)
+{
+  if(target->is_function_call())
+  {
+    if(target->call_function().id() == ID_symbol)
+    {
+      // this is a function call
+      if(do_not_instrument(
+           to_symbol_expr(target->call_function()).get_identifier()))
+      {
+        // do not pass write set if we know the function is not instrumented
+        return;
+      }
+      // pass write set argument
+      target->call_arguments().push_back(write_set);
```

Comment:
Perhaps negate the condition of the `if` statement, get rid of `return` and move this instruction into the body?

Reply:
fixed